samples/dpdk_app/dpdk-app-centos/README.md

```diff
@@ -0,0 +1,201 @@
+DPDK running in a container.
+
+## SR-IOV CNI and SR-IOV Device Plugin
+The SR-IOV Device Plugin detects and tracks the VFs associated with an
+SR-IOV PF. The SR-IOV CNI updates the VF as needed and the PCI Address
+associated woth the VF is passed tp the container via environmental
+variables. The container, like this one, boots up and reads the
+environmental variables (via the app-netuilt) and runs a DPDK application.
+
+## Userspace CNI
+The Userspace CNI inconjunction with the OVS CNI Library (cniovs) or VPP
+CNI Library (cnivpp) creates interfaces on the host, like a vhost-user or
+a memif interface, adds the host side of the interface to a local network,
+like a L2 bridge, then copies information needed in the container into
+annotations. The container, like this one, boots up and reads the
+annotatations (via the app-netuilt) and runs a DPDK application.
+
+
+# Build Instructions for dpdk-app-centos Docker Image
+Get the **app-netutil** repo:
+```
+   cd $GOPATH/src/
+   go get github.com/openshift/app-netutil
+```
+
+Build the docker image:
+```
+   cd $GOPATH/src/github.com/openshift/app-netutil/samples/dpdk_app/dpdk-app-centos/
+   docker build --rm -t dpdk-app-centos .
+```
+
+## Reduce Image Size
+Multi-stage builds are a new feature requiring **Docker 17.05** or higher on
+the daemon and client. If multi-stage builds are supported on your system,
+then uncomment the following lines (those with **##**) in the Dockerfile:
+```
+:
+
+# -------- Import stage.
+# Docker 17.05 or higher
+##FROM centos
+##COPY --from=0 /usr/bin/dpdk-app /usr/bin/dpdk-app
+##COPY --from=0 /usr/bin/l3fwd /usr/bin/l3fwd
+##COPY --from=0 /usr/bin/testpmd /usr/bin/testpmd
+##COPY --from=0 /lib64/libnetutil_api.so /lib64/libnetutil_api.so
+##COPY --from=0 /usr/lib64/libnuma.so.1 /usr/lib64/libnuma.so.1
+
+:
+```
+
+# Docker Image Details
+This Docker Image is downloading DPDK (version 19.08 to get memif PMD)
+and building it. Once built, it changes into the DPDK `testpmd`
+directory (${DPDK_DIR}/app/test-pmd) and builds it. It then changes
+in the DPDK `l3fwd` directory (${DPDK_DIR}/examples/l3fwd) and builds
+it.
+
+`testpmd` and `l3fwd` are DPDK sample applications. `testpmd` performs
+Layer 3 switching and `l3fwd` performs Layer 3 routing. Both applications
+are built with the `app-netutil` and copied into '/usr/bin/'. One of the
+applications is then also copied to '/usr/bin/' and renamed as `dpdk-app`.
+See your Dockerfile, because multiple versions exist and copy different
+applications as `dpdk-app`. Below assumes `l3fwd` is copied over as
+`dpdk-app`. But same concept applies to `testpmd`.
+
+Typically, `l3fwd` is started with a set of input parameters that initializes DPDK.
+For example:
+```
+$ l3fwd -n 4 -l 1 --master-lcore 1 -w 0000:01:0a.6 -w 0000:01:02.1 -- -p 0x3 -P --config="(0,0,1),(1,0,1)" --parse-ptype
+```
+
+This Docker image is tweaking this a little. Before `l3fwd` is built, the
```

Comment:
Thanks Billy for writing this sample! I cannot believe that it's changing the source code with sed automatically. That should have required a mount of work :+1:

Reply:
agree, truly nice work!